rift/node.py

```diff
@@ -2560,49 +2637,63 @@ def spf_consider_candidate_dest(self, destination, nbr_tie_element, predecessor_
                 # The new path is strictly better than the existing path. Replace the existing path
                 # with the new path.
                 self.set_spf_predecessor(destination, nbr_tie_element, predecessor_system_id,
-                                         spf_direction)
+                                         spf_direction, special_for_neg_disagg)
+
                 dest_table[dest_key] = destination
                 candidates[dest_key] = destination.cost
             elif destination.cost == old_destination.cost:
                 # The new path is equal cost to the existing path. Add an ECMP path to the existing
                 # path.
-                self.add_spf_predecessor(old_destination, predecessor_system_id, spf_direction)
+                self.add_spf_predecessor(old_destination, predecessor_system_id,
+                                         spf_direction, special_for_neg_disagg)
+
                 old_destination.inherit_tags(destination)
 
     def set_spf_predecessor(self, destination, nbr_tie_element, predecessor_system_id,
-                            spf_direction):
+                            spf_direction, special_for_neg_disagg):
         destination.add_predecessor(predecessor_system_id)
+
         if (nbr_tie_element is not None) and (predecessor_system_id == self.system_id):
             for link_id_pair in nbr_tie_element.link_ids:
                 nhop = self.interface_id_to_ipv4_next_hop(link_id_pair.local_id)
                 if nhop:
                     destination.add_ipv4_next_hop(nhop)
+
                 nhop = self.interface_id_to_ipv6_next_hop(link_id_pair.local_id)
                 if nhop:
                     destination.add_ipv6_next_hop(nhop)
         else:
-            dest_table = self._spf_destinations[spf_direction]
+            dest_table = self._spf_destinations[(spf_direction, special_for_neg_disagg)]
             destination.inherit_next_hops(dest_table[predecessor_system_id])
 
-    def add_spf_predecessor(self, destination, predecessor_system_id, spf_direction):
+    def add_spf_predecessor(self, destination, predecessor_system_id, spf_direction, special_for_neg_disagg):
         destination.add_predecessor(predecessor_system_id)
-        dest_table = self._spf_destinations[spf_direction]
+        dest_table = self._spf_destinations[(spf_direction, special_for_neg_disagg)]
         destination.inherit_next_hops(dest_table[predecessor_system_id])
 
     def spf_mark_pos_disagg_prefixes(self):
         # Mark the prefixes in the SPF table for which this router wants to do positive aggregation
         # (not to be confused with prefixes in this SPF tables which were received because some
         # north-bound router did positive disaggregation)
-        for dest in self._spf_destinations[constants.DIR_SOUTH].values():
+
+        # FIXME: Prefilter only dests with dest_type == spf_dest.DEST_TYPE_PREFIX?
```

Comment:
@brunorijsman can we use a more Pythonic approach here?

Reply:
Personally I find the code I have (without the filter) easier to read, but I admit it is a matter of personal preference (also easier to debug with breakpoints!).